Ejecutando con 5000000 elementos
Time taken by reduction operation using a function: 2.6 ms Â± 206 ns per loop (mean Â± std. dev. of 2 runs, 100 loops each)
And the result of the sum of numbers in the range [0, value) is: 24962.640155252877

Time taken by reduction operation using numpy.sum(): 7.93 Âµs Â± 20.2 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.sum(): 24962.64015525305 
 
Time taken by reduction operation using numpy.ndarray.sum(): 7.12 Âµs Â± 2.05 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.ndarray.sum(): 24962.64015525305
[0;31m---------------------------------------------------------------------------[0m
[0;31mOSError[0m                                   Traceback (most recent call last)
File [0;32mcupy_backends/cuda/_softlink.pyx:25[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/ctypes/__init__.py:374[0m, in [0;36mCDLL.__init__[0;34m(self, name, mode, handle, use_errno, use_last_error, winmode)[0m
[1;32m    373[0m [38;5;28;01mif[39;00m handle [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 374[0m     [38;5;28mself[39m[38;5;241m.[39m_handle [38;5;241m=[39m [43m_dlopen[49m[43m([49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_name[49m[43m,[49m[43m [49m[43mmode[49m[43m)[49m
[1;32m    375[0m [38;5;28;01melse[39;00m:

[0;31mOSError[0m: libnvrtc.so.12: cannot open shared object file: No such file or directory

The above exception was the direct cause of the following exception:

[0;31mRuntimeError[0m                              Traceback (most recent call last)
Cell [0;32mIn[1], line 14[0m
[1;32m     11[0m value [38;5;241m=[39m [38;5;241m5[39m [38;5;241m*[39m [38;5;241m10[39m[38;5;241m*[39m[38;5;241m*[39m[38;5;241m4[39m  [38;5;66;03m# TamaÃ±o del array[39;00m
[1;32m     13[0m [38;5;66;03m# Crear el array directamente en la GPU utilizando cupy.random.rand[39;00m
[0;32m---> 14[0m X_gpu [38;5;241m=[39m [43mcp[49m[38;5;241;43m.[39;49m[43mrandom[49m[38;5;241;43m.[39;49m[43mrand[49m[43m([49m[43mvalue[49m[43m)[49m  
[1;32m     16[0m [38;5;66;03m# Para imprimir los primeros valores del array (traspasÃ¡ndolos a la CPU con .get())[39;00m
[1;32m     17[0m [38;5;66;03m# print(cp.asnumpy(X_gpu[:12]))[39;00m
[1;32m     18[0m 
[1;32m     19[0m [38;5;66;03m# Reducir operaciÃ³n en la GPU con cupy.sum()[39;00m
[1;32m     20[0m tiempo [38;5;241m=[39m get_ipython()[38;5;241m.[39mrun_line_magic([38;5;124m'[39m[38;5;124mtimeit[39m[38;5;124m'[39m, [38;5;124m'[39m[38;5;124m-r 2 -o -q cp.sum(X_gpu)[39m[38;5;124m'[39m)

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:44[0m, in [0;36mrand[0;34m(*size, **kwarg)[0m
[1;32m     41[0m [38;5;28;01mif[39;00m kwarg:
[1;32m     42[0m     [38;5;28;01mraise[39;00m [38;5;167;01mTypeError[39;00m([38;5;124m'[39m[38;5;124mrand() got unexpected keyword arguments [39m[38;5;132;01m%s[39;00m[38;5;124m'[39m
[1;32m     43[0m                     [38;5;241m%[39m [38;5;124m'[39m[38;5;124m, [39m[38;5;124m'[39m[38;5;241m.[39mjoin(kwarg[38;5;241m.[39mkeys()))
[0;32m---> 44[0m [38;5;28;01mreturn[39;00m [43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:156[0m, in [0;36mrandom_sample[0;34m(size, dtype)[0m
[1;32m    140[0m [38;5;124;03m"""Returns an array of random values over the interval ``[0, 1)``.[39;00m
[1;32m    141[0m 
[1;32m    142[0m [38;5;124;03mThis is a variant of :func:`cupy.random.rand`.[39;00m
[0;32m   (...)[0m
[1;32m    153[0m 
[1;32m    154[0m [38;5;124;03m"""[39;00m
[1;32m    155[0m rs [38;5;241m=[39m _generator[38;5;241m.[39mget_random_state()
[0;32m--> 156[0m [38;5;28;01mreturn[39;00m [43mrs[49m[38;5;241;43m.[39;49m[43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_generator.py:630[0m, in [0;36mRandomState.random_sample[0;34m(self, size, dtype)[0m
[1;32m    628[0m     size [38;5;241m=[39m ()
[1;32m    629[0m out [38;5;241m=[39m [38;5;28mself[39m[38;5;241m.[39m_random_sample_raw(size, dtype)
[0;32m--> 630[0m [43mRandomState[49m[38;5;241;43m.[39;49m[43m_mod1_kernel[49m[43m([49m[43mout[49m[43m)[49m
[1;32m    631[0m [38;5;28;01mreturn[39;00m out

File [0;32mcupy/_core/_kernel.pyx:920[0m, in [0;36mcupy._core._kernel.ElementwiseKernel.__call__[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:945[0m, in [0;36mcupy._core._kernel.ElementwiseKernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:728[0m, in [0;36mcupy._core._kernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:82[0m, in [0;36mcupy._core._kernel._get_simple_elementwise_kernel_from_code[0;34m()[0m

File [0;32mcupy/_core/core.pyx:2254[0m, in [0;36mcupy._core.core.compile_with_cache[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:484[0m, in [0;36m_compile_module_with_cache[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, jitify)[0m
[1;32m    480[0m     [38;5;28;01mreturn[39;00m _compile_with_cache_hip(
[1;32m    481[0m         source, options, arch, cache_dir, extra_source, backend,
[1;32m    482[0m         name_expressions, log_stream, cache_in_memory)
[1;32m    483[0m [38;5;28;01melse[39;00m:
[0;32m--> 484[0m     [38;5;28;01mreturn[39;00m [43m_compile_with_cache_cuda[49m[43m([49m
[1;32m    485[0m [43m        [49m[43msource[49m[43m,[49m[43m [49m[43moptions[49m[43m,[49m[43m [49m[43march[49m[43m,[49m[43m [49m[43mcache_dir[49m[43m,[49m[43m [49m[43mextra_source[49m[43m,[49m[43m [49m[43mbackend[49m[43m,[49m
[1;32m    486[0m [43m        [49m[43menable_cooperative_groups[49m[43m,[49m[43m [49m[43mname_expressions[49m[43m,[49m[43m [49m[43mlog_stream[49m[43m,[49m
[1;32m    487[0m [43m        [49m[43mcache_in_memory[49m[43m,[49m[43m [49m[43mjitify[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:499[0m, in [0;36m_compile_with_cache_cuda[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory, jitify)[0m
[1;32m    497[0m     cache_dir [38;5;241m=[39m get_cache_dir()
[1;32m    498[0m [38;5;28;01mif[39;00m arch [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 499[0m     arch [38;5;241m=[39m [43m_get_arch[49m[43m([49m[43m)[49m
[1;32m    501[0m options [38;5;241m+[39m[38;5;241m=[39m ([38;5;124m'[39m[38;5;124m-ftz=true[39m[38;5;124m'[39m,)
[1;32m    503[0m [38;5;28;01mif[39;00m enable_cooperative_groups:
[1;32m    504[0m     [38;5;66;03m# `cooperative_groups` requires relocatable device code.[39;00m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:148[0m, in [0;36m_get_arch[0;34m()[0m
[1;32m    144[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize(for_each_device[38;5;241m=[39m[38;5;28;01mTrue[39;00m)
[1;32m    145[0m [38;5;28;01mdef[39;00m [38;5;21m_get_arch[39m():
[1;32m    146[0m     [38;5;66;03m# See Supported Compile Options section of NVRTC User Guide for[39;00m
[1;32m    147[0m     [38;5;66;03m# the maximum value allowed for `--gpu-architecture`.[39;00m
[0;32m--> 148[0m     nvrtc_max_compute_capability [38;5;241m=[39m [43m_get_max_compute_capability[49m[43m([49m[43m)[49m
[1;32m    150[0m     arch [38;5;241m=[39m device[38;5;241m.[39mDevice()[38;5;241m.[39mcompute_capability
[1;32m    151[0m     [38;5;28;01mif[39;00m arch [38;5;129;01min[39;00m _tegra_archs:

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:126[0m, in [0;36m_get_max_compute_capability[0;34m()[0m
[1;32m    124[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize()
[1;32m    125[0m [38;5;28;01mdef[39;00m [38;5;21m_get_max_compute_capability[39m():
[0;32m--> 126[0m     major, minor [38;5;241m=[39m [43m_get_nvrtc_version[49m[43m([49m[43m)[49m
[1;32m    127[0m     [38;5;28;01mif[39;00m major [38;5;241m<[39m [38;5;241m11[39m:
[1;32m    128[0m         [38;5;66;03m# CUDA 10.2[39;00m
[1;32m    129[0m         nvrtc_max_compute_capability [38;5;241m=[39m [38;5;124m'[39m[38;5;124m75[39m[38;5;124m'[39m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:115[0m, in [0;36m_get_nvrtc_version[0;34m()[0m
[1;32m    113[0m [38;5;28;01mglobal[39;00m _nvrtc_version
[1;32m    114[0m [38;5;28;01mif[39;00m _nvrtc_version [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 115[0m     _nvrtc_version [38;5;241m=[39m [43mnvrtc[49m[38;5;241;43m.[39;49m[43mgetVersion[49m[43m([49m[43m)[49m
[1;32m    117[0m [38;5;28;01mreturn[39;00m _nvrtc_version

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:56[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:57[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:72[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:76[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:143[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._get_softlink[0;34m()[0m

File [0;32mcupy_backends/cuda/_softlink.pyx:32[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

[0;31mRuntimeError[0m: CuPy failed to load libnvrtc.so.12: OSError: libnvrtc.so.12: cannot open shared object file: No such file or directory
Ejecutando con 50000000 elementos
Time taken by reduction operation using a function: 2.58 ms Â± 2.61 Âµs per loop (mean Â± std. dev. of 2 runs, 100 loops each)
And the result of the sum of numbers in the range [0, value) is: 25015.251304754995

Time taken by reduction operation using numpy.sum(): 7.97 Âµs Â± 138 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.sum(): 25015.25130475496 
 
Time taken by reduction operation using numpy.ndarray.sum(): 7.02 Âµs Â± 1.24 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.ndarray.sum(): 25015.25130475496
[0;31m---------------------------------------------------------------------------[0m
[0;31mOSError[0m                                   Traceback (most recent call last)
File [0;32mcupy_backends/cuda/_softlink.pyx:25[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/ctypes/__init__.py:374[0m, in [0;36mCDLL.__init__[0;34m(self, name, mode, handle, use_errno, use_last_error, winmode)[0m
[1;32m    373[0m [38;5;28;01mif[39;00m handle [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 374[0m     [38;5;28mself[39m[38;5;241m.[39m_handle [38;5;241m=[39m [43m_dlopen[49m[43m([49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_name[49m[43m,[49m[43m [49m[43mmode[49m[43m)[49m
[1;32m    375[0m [38;5;28;01melse[39;00m:

[0;31mOSError[0m: libnvrtc.so.12: cannot open shared object file: No such file or directory

The above exception was the direct cause of the following exception:

[0;31mRuntimeError[0m                              Traceback (most recent call last)
Cell [0;32mIn[1], line 14[0m
[1;32m     11[0m value [38;5;241m=[39m [38;5;241m5[39m [38;5;241m*[39m [38;5;241m10[39m[38;5;241m*[39m[38;5;241m*[39m[38;5;241m4[39m  [38;5;66;03m# TamaÃ±o del array[39;00m
[1;32m     13[0m [38;5;66;03m# Crear el array directamente en la GPU utilizando cupy.random.rand[39;00m
[0;32m---> 14[0m X_gpu [38;5;241m=[39m [43mcp[49m[38;5;241;43m.[39;49m[43mrandom[49m[38;5;241;43m.[39;49m[43mrand[49m[43m([49m[43mvalue[49m[43m)[49m  
[1;32m     16[0m [38;5;66;03m# Para imprimir los primeros valores del array (traspasÃ¡ndolos a la CPU con .get())[39;00m
[1;32m     17[0m [38;5;66;03m# print(cp.asnumpy(X_gpu[:12]))[39;00m
[1;32m     18[0m 
[1;32m     19[0m [38;5;66;03m# Reducir operaciÃ³n en la GPU con cupy.sum()[39;00m
[1;32m     20[0m tiempo [38;5;241m=[39m get_ipython()[38;5;241m.[39mrun_line_magic([38;5;124m'[39m[38;5;124mtimeit[39m[38;5;124m'[39m, [38;5;124m'[39m[38;5;124m-r 2 -o -q cp.sum(X_gpu)[39m[38;5;124m'[39m)

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:44[0m, in [0;36mrand[0;34m(*size, **kwarg)[0m
[1;32m     41[0m [38;5;28;01mif[39;00m kwarg:
[1;32m     42[0m     [38;5;28;01mraise[39;00m [38;5;167;01mTypeError[39;00m([38;5;124m'[39m[38;5;124mrand() got unexpected keyword arguments [39m[38;5;132;01m%s[39;00m[38;5;124m'[39m
[1;32m     43[0m                     [38;5;241m%[39m [38;5;124m'[39m[38;5;124m, [39m[38;5;124m'[39m[38;5;241m.[39mjoin(kwarg[38;5;241m.[39mkeys()))
[0;32m---> 44[0m [38;5;28;01mreturn[39;00m [43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:156[0m, in [0;36mrandom_sample[0;34m(size, dtype)[0m
[1;32m    140[0m [38;5;124;03m"""Returns an array of random values over the interval ``[0, 1)``.[39;00m
[1;32m    141[0m 
[1;32m    142[0m [38;5;124;03mThis is a variant of :func:`cupy.random.rand`.[39;00m
[0;32m   (...)[0m
[1;32m    153[0m 
[1;32m    154[0m [38;5;124;03m"""[39;00m
[1;32m    155[0m rs [38;5;241m=[39m _generator[38;5;241m.[39mget_random_state()
[0;32m--> 156[0m [38;5;28;01mreturn[39;00m [43mrs[49m[38;5;241;43m.[39;49m[43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_generator.py:630[0m, in [0;36mRandomState.random_sample[0;34m(self, size, dtype)[0m
[1;32m    628[0m     size [38;5;241m=[39m ()
[1;32m    629[0m out [38;5;241m=[39m [38;5;28mself[39m[38;5;241m.[39m_random_sample_raw(size, dtype)
[0;32m--> 630[0m [43mRandomState[49m[38;5;241;43m.[39;49m[43m_mod1_kernel[49m[43m([49m[43mout[49m[43m)[49m
[1;32m    631[0m [38;5;28;01mreturn[39;00m out

File [0;32mcupy/_core/_kernel.pyx:920[0m, in [0;36mcupy._core._kernel.ElementwiseKernel.__call__[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:945[0m, in [0;36mcupy._core._kernel.ElementwiseKernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:728[0m, in [0;36mcupy._core._kernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:82[0m, in [0;36mcupy._core._kernel._get_simple_elementwise_kernel_from_code[0;34m()[0m

File [0;32mcupy/_core/core.pyx:2254[0m, in [0;36mcupy._core.core.compile_with_cache[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:484[0m, in [0;36m_compile_module_with_cache[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, jitify)[0m
[1;32m    480[0m     [38;5;28;01mreturn[39;00m _compile_with_cache_hip(
[1;32m    481[0m         source, options, arch, cache_dir, extra_source, backend,
[1;32m    482[0m         name_expressions, log_stream, cache_in_memory)
[1;32m    483[0m [38;5;28;01melse[39;00m:
[0;32m--> 484[0m     [38;5;28;01mreturn[39;00m [43m_compile_with_cache_cuda[49m[43m([49m
[1;32m    485[0m [43m        [49m[43msource[49m[43m,[49m[43m [49m[43moptions[49m[43m,[49m[43m [49m[43march[49m[43m,[49m[43m [49m[43mcache_dir[49m[43m,[49m[43m [49m[43mextra_source[49m[43m,[49m[43m [49m[43mbackend[49m[43m,[49m
[1;32m    486[0m [43m        [49m[43menable_cooperative_groups[49m[43m,[49m[43m [49m[43mname_expressions[49m[43m,[49m[43m [49m[43mlog_stream[49m[43m,[49m
[1;32m    487[0m [43m        [49m[43mcache_in_memory[49m[43m,[49m[43m [49m[43mjitify[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:499[0m, in [0;36m_compile_with_cache_cuda[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory, jitify)[0m
[1;32m    497[0m     cache_dir [38;5;241m=[39m get_cache_dir()
[1;32m    498[0m [38;5;28;01mif[39;00m arch [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 499[0m     arch [38;5;241m=[39m [43m_get_arch[49m[43m([49m[43m)[49m
[1;32m    501[0m options [38;5;241m+[39m[38;5;241m=[39m ([38;5;124m'[39m[38;5;124m-ftz=true[39m[38;5;124m'[39m,)
[1;32m    503[0m [38;5;28;01mif[39;00m enable_cooperative_groups:
[1;32m    504[0m     [38;5;66;03m# `cooperative_groups` requires relocatable device code.[39;00m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:148[0m, in [0;36m_get_arch[0;34m()[0m
[1;32m    144[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize(for_each_device[38;5;241m=[39m[38;5;28;01mTrue[39;00m)
[1;32m    145[0m [38;5;28;01mdef[39;00m [38;5;21m_get_arch[39m():
[1;32m    146[0m     [38;5;66;03m# See Supported Compile Options section of NVRTC User Guide for[39;00m
[1;32m    147[0m     [38;5;66;03m# the maximum value allowed for `--gpu-architecture`.[39;00m
[0;32m--> 148[0m     nvrtc_max_compute_capability [38;5;241m=[39m [43m_get_max_compute_capability[49m[43m([49m[43m)[49m
[1;32m    150[0m     arch [38;5;241m=[39m device[38;5;241m.[39mDevice()[38;5;241m.[39mcompute_capability
[1;32m    151[0m     [38;5;28;01mif[39;00m arch [38;5;129;01min[39;00m _tegra_archs:

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:126[0m, in [0;36m_get_max_compute_capability[0;34m()[0m
[1;32m    124[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize()
[1;32m    125[0m [38;5;28;01mdef[39;00m [38;5;21m_get_max_compute_capability[39m():
[0;32m--> 126[0m     major, minor [38;5;241m=[39m [43m_get_nvrtc_version[49m[43m([49m[43m)[49m
[1;32m    127[0m     [38;5;28;01mif[39;00m major [38;5;241m<[39m [38;5;241m11[39m:
[1;32m    128[0m         [38;5;66;03m# CUDA 10.2[39;00m
[1;32m    129[0m         nvrtc_max_compute_capability [38;5;241m=[39m [38;5;124m'[39m[38;5;124m75[39m[38;5;124m'[39m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:115[0m, in [0;36m_get_nvrtc_version[0;34m()[0m
[1;32m    113[0m [38;5;28;01mglobal[39;00m _nvrtc_version
[1;32m    114[0m [38;5;28;01mif[39;00m _nvrtc_version [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 115[0m     _nvrtc_version [38;5;241m=[39m [43mnvrtc[49m[38;5;241;43m.[39;49m[43mgetVersion[49m[43m([49m[43m)[49m
[1;32m    117[0m [38;5;28;01mreturn[39;00m _nvrtc_version

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:56[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:57[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:72[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:76[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:143[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._get_softlink[0;34m()[0m

File [0;32mcupy_backends/cuda/_softlink.pyx:32[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

[0;31mRuntimeError[0m: CuPy failed to load libnvrtc.so.12: OSError: libnvrtc.so.12: cannot open shared object file: No such file or directory
Ejecutando con 500000000 elementos
Time taken by reduction operation using a function: 2.62 ms Â± 4.14 Âµs per loop (mean Â± std. dev. of 2 runs, 100 loops each)
And the result of the sum of numbers in the range [0, value) is: 25046.45897359304

Time taken by reduction operation using numpy.sum(): 7.77 Âµs Â± 0.829 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.sum(): 25046.458973593148 
 
Time taken by reduction operation using numpy.ndarray.sum(): 7 Âµs Â± 4.35 ns per loop (mean Â± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.ndarray.sum(): 25046.458973593148
[0;31m---------------------------------------------------------------------------[0m
[0;31mOSError[0m                                   Traceback (most recent call last)
File [0;32mcupy_backends/cuda/_softlink.pyx:25[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/ctypes/__init__.py:374[0m, in [0;36mCDLL.__init__[0;34m(self, name, mode, handle, use_errno, use_last_error, winmode)[0m
[1;32m    373[0m [38;5;28;01mif[39;00m handle [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 374[0m     [38;5;28mself[39m[38;5;241m.[39m_handle [38;5;241m=[39m [43m_dlopen[49m[43m([49m[38;5;28;43mself[39;49m[38;5;241;43m.[39;49m[43m_name[49m[43m,[49m[43m [49m[43mmode[49m[43m)[49m
[1;32m    375[0m [38;5;28;01melse[39;00m:

[0;31mOSError[0m: libnvrtc.so.12: cannot open shared object file: No such file or directory

The above exception was the direct cause of the following exception:

[0;31mRuntimeError[0m                              Traceback (most recent call last)
Cell [0;32mIn[1], line 14[0m
[1;32m     11[0m value [38;5;241m=[39m [38;5;241m5[39m [38;5;241m*[39m [38;5;241m10[39m[38;5;241m*[39m[38;5;241m*[39m[38;5;241m4[39m  [38;5;66;03m# TamaÃ±o del array[39;00m
[1;32m     13[0m [38;5;66;03m# Crear el array directamente en la GPU utilizando cupy.random.rand[39;00m
[0;32m---> 14[0m X_gpu [38;5;241m=[39m [43mcp[49m[38;5;241;43m.[39;49m[43mrandom[49m[38;5;241;43m.[39;49m[43mrand[49m[43m([49m[43mvalue[49m[43m)[49m  
[1;32m     16[0m [38;5;66;03m# Para imprimir los primeros valores del array (traspasÃ¡ndolos a la CPU con .get())[39;00m
[1;32m     17[0m [38;5;66;03m# print(cp.asnumpy(X_gpu[:12]))[39;00m
[1;32m     18[0m 
[1;32m     19[0m [38;5;66;03m# Reducir operaciÃ³n en la GPU con cupy.sum()[39;00m
[1;32m     20[0m tiempo [38;5;241m=[39m get_ipython()[38;5;241m.[39mrun_line_magic([38;5;124m'[39m[38;5;124mtimeit[39m[38;5;124m'[39m, [38;5;124m'[39m[38;5;124m-r 2 -o -q cp.sum(X_gpu)[39m[38;5;124m'[39m)

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:44[0m, in [0;36mrand[0;34m(*size, **kwarg)[0m
[1;32m     41[0m [38;5;28;01mif[39;00m kwarg:
[1;32m     42[0m     [38;5;28;01mraise[39;00m [38;5;167;01mTypeError[39;00m([38;5;124m'[39m[38;5;124mrand() got unexpected keyword arguments [39m[38;5;132;01m%s[39;00m[38;5;124m'[39m
[1;32m     43[0m                     [38;5;241m%[39m [38;5;124m'[39m[38;5;124m, [39m[38;5;124m'[39m[38;5;241m.[39mjoin(kwarg[38;5;241m.[39mkeys()))
[0;32m---> 44[0m [38;5;28;01mreturn[39;00m [43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_sample.py:156[0m, in [0;36mrandom_sample[0;34m(size, dtype)[0m
[1;32m    140[0m [38;5;124;03m"""Returns an array of random values over the interval ``[0, 1)``.[39;00m
[1;32m    141[0m 
[1;32m    142[0m [38;5;124;03mThis is a variant of :func:`cupy.random.rand`.[39;00m
[0;32m   (...)[0m
[1;32m    153[0m 
[1;32m    154[0m [38;5;124;03m"""[39;00m
[1;32m    155[0m rs [38;5;241m=[39m _generator[38;5;241m.[39mget_random_state()
[0;32m--> 156[0m [38;5;28;01mreturn[39;00m [43mrs[49m[38;5;241;43m.[39;49m[43mrandom_sample[49m[43m([49m[43msize[49m[38;5;241;43m=[39;49m[43msize[49m[43m,[49m[43m [49m[43mdtype[49m[38;5;241;43m=[39;49m[43mdtype[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/random/_generator.py:630[0m, in [0;36mRandomState.random_sample[0;34m(self, size, dtype)[0m
[1;32m    628[0m     size [38;5;241m=[39m ()
[1;32m    629[0m out [38;5;241m=[39m [38;5;28mself[39m[38;5;241m.[39m_random_sample_raw(size, dtype)
[0;32m--> 630[0m [43mRandomState[49m[38;5;241;43m.[39;49m[43m_mod1_kernel[49m[43m([49m[43mout[49m[43m)[49m
[1;32m    631[0m [38;5;28;01mreturn[39;00m out

File [0;32mcupy/_core/_kernel.pyx:920[0m, in [0;36mcupy._core._kernel.ElementwiseKernel.__call__[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:945[0m, in [0;36mcupy._core._kernel.ElementwiseKernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:728[0m, in [0;36mcupy._core._kernel._get_elementwise_kernel[0;34m()[0m

File [0;32mcupy/_core/_kernel.pyx:82[0m, in [0;36mcupy._core._kernel._get_simple_elementwise_kernel_from_code[0;34m()[0m

File [0;32mcupy/_core/core.pyx:2254[0m, in [0;36mcupy._core.core.compile_with_cache[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:484[0m, in [0;36m_compile_module_with_cache[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, jitify)[0m
[1;32m    480[0m     [38;5;28;01mreturn[39;00m _compile_with_cache_hip(
[1;32m    481[0m         source, options, arch, cache_dir, extra_source, backend,
[1;32m    482[0m         name_expressions, log_stream, cache_in_memory)
[1;32m    483[0m [38;5;28;01melse[39;00m:
[0;32m--> 484[0m     [38;5;28;01mreturn[39;00m [43m_compile_with_cache_cuda[49m[43m([49m
[1;32m    485[0m [43m        [49m[43msource[49m[43m,[49m[43m [49m[43moptions[49m[43m,[49m[43m [49m[43march[49m[43m,[49m[43m [49m[43mcache_dir[49m[43m,[49m[43m [49m[43mextra_source[49m[43m,[49m[43m [49m[43mbackend[49m[43m,[49m
[1;32m    486[0m [43m        [49m[43menable_cooperative_groups[49m[43m,[49m[43m [49m[43mname_expressions[49m[43m,[49m[43m [49m[43mlog_stream[49m[43m,[49m
[1;32m    487[0m [43m        [49m[43mcache_in_memory[49m[43m,[49m[43m [49m[43mjitify[49m[43m)[49m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:499[0m, in [0;36m_compile_with_cache_cuda[0;34m(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory, jitify)[0m
[1;32m    497[0m     cache_dir [38;5;241m=[39m get_cache_dir()
[1;32m    498[0m [38;5;28;01mif[39;00m arch [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 499[0m     arch [38;5;241m=[39m [43m_get_arch[49m[43m([49m[43m)[49m
[1;32m    501[0m options [38;5;241m+[39m[38;5;241m=[39m ([38;5;124m'[39m[38;5;124m-ftz=true[39m[38;5;124m'[39m,)
[1;32m    503[0m [38;5;28;01mif[39;00m enable_cooperative_groups:
[1;32m    504[0m     [38;5;66;03m# `cooperative_groups` requires relocatable device code.[39;00m

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:148[0m, in [0;36m_get_arch[0;34m()[0m
[1;32m    144[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize(for_each_device[38;5;241m=[39m[38;5;28;01mTrue[39;00m)
[1;32m    145[0m [38;5;28;01mdef[39;00m [38;5;21m_get_arch[39m():
[1;32m    146[0m     [38;5;66;03m# See Supported Compile Options section of NVRTC User Guide for[39;00m
[1;32m    147[0m     [38;5;66;03m# the maximum value allowed for `--gpu-architecture`.[39;00m
[0;32m--> 148[0m     nvrtc_max_compute_capability [38;5;241m=[39m [43m_get_max_compute_capability[49m[43m([49m[43m)[49m
[1;32m    150[0m     arch [38;5;241m=[39m device[38;5;241m.[39mDevice()[38;5;241m.[39mcompute_capability
[1;32m    151[0m     [38;5;28;01mif[39;00m arch [38;5;129;01min[39;00m _tegra_archs:

File [0;32mcupy/_util.pyx:64[0m, in [0;36mcupy._util.memoize.decorator.ret[0;34m()[0m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:126[0m, in [0;36m_get_max_compute_capability[0;34m()[0m
[1;32m    124[0m [38;5;129m@_util[39m[38;5;241m.[39mmemoize()
[1;32m    125[0m [38;5;28;01mdef[39;00m [38;5;21m_get_max_compute_capability[39m():
[0;32m--> 126[0m     major, minor [38;5;241m=[39m [43m_get_nvrtc_version[49m[43m([49m[43m)[49m
[1;32m    127[0m     [38;5;28;01mif[39;00m major [38;5;241m<[39m [38;5;241m11[39m:
[1;32m    128[0m         [38;5;66;03m# CUDA 10.2[39;00m
[1;32m    129[0m         nvrtc_max_compute_capability [38;5;241m=[39m [38;5;124m'[39m[38;5;124m75[39m[38;5;124m'[39m

File [0;32m/home/modules/anaconda3-2023/lib/python3.10/site-packages/cupy/cuda/compiler.py:115[0m, in [0;36m_get_nvrtc_version[0;34m()[0m
[1;32m    113[0m [38;5;28;01mglobal[39;00m _nvrtc_version
[1;32m    114[0m [38;5;28;01mif[39;00m _nvrtc_version [38;5;129;01mis[39;00m [38;5;28;01mNone[39;00m:
[0;32m--> 115[0m     _nvrtc_version [38;5;241m=[39m [43mnvrtc[49m[38;5;241;43m.[39;49m[43mgetVersion[49m[43m([49m[43m)[49m
[1;32m    117[0m [38;5;28;01mreturn[39;00m _nvrtc_version

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:56[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/nvrtc.pyx:57[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.getVersion[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:72[0m, in [0;36mcupy_backends.cuda.libs.nvrtc.initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:76[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._initialize[0;34m()[0m

File [0;32mcupy_backends/cuda/libs/_cnvrtc.pxi:143[0m, in [0;36mcupy_backends.cuda.libs.nvrtc._get_softlink[0;34m()[0m

File [0;32mcupy_backends/cuda/_softlink.pyx:32[0m, in [0;36mcupy_backends.cuda._softlink.SoftLink.__init__[0;34m()[0m

[0;31mRuntimeError[0m: CuPy failed to load libnvrtc.so.12: OSError: libnvrtc.so.12: cannot open shared object file: No such file or directory
